test/thread_safety_test.rb

```diff
@@ -0,0 +1,273 @@
+      max_parallel_time = minimum_serial_query_exection_time - 0.1
+      assert_operator(max_parallel_time, :>, thread_exection_time)
+
+      # This order cannot be guaranteed but it likely given the artifical delays
+      rows = results.map(&:first)
+      result_strings = rows.map { |r| r[1] }
+      expect(result_strings).must_equal(
+        [
+          "faster query",
+          "medium query",
+          "slower query"
+        ]
+      )
+    end
+  end
+
+  describe "when multiple threads use the same database" do
+    it "exposes a different connections to each thread" do
+      connections = []
+
+      new_thread("connection_per_thread1") do
+        ActiveRecord::Base.on_primary do
+          pause_and_mark_ready
+          connections << ActiveRecord::Base.connection
+        end
+      end
+
+      new_thread("connection_per_thread2") do
+        ActiveRecord::Base.on_primary do
+          pause_and_mark_ready
+          connections << ActiveRecord::Base.connection
+        end
+      end
+
+      wait_for_threads_to_be_ready
+      execute_and_wait_for_threads
+
+      expect(connections.first).must_be_kind_of(ActiveRecord::ConnectionAdapters::Mysql2Adapter)
+      expect(connections.uniq.size).must_equal(2)
+    end
+
+    it "allows threads to parallelize their IO" do
+      results = []
+
+      query_delay = { fast: "0.01", slow: "1", medium: "0.5" }
+      new_thread("same_db_parallel_thread1") do
+        ActiveRecord::Base.on_primary do
+          pause_and_mark_ready
+          result = execute_sql("SELECT 'slower query',SLEEP(#{query_delay.fetch(:slow)})")
+          results.push(result)
+        end
+      end
+
+      new_thread("same_db_parallel_thread2") do
+        ActiveRecord::Base.on_primary do
+          pause_and_mark_ready
+          result = execute_sql("SELECT 'faster query',SLEEP(#{query_delay.fetch(:fast)})")
+          results.push(result)
+        end
+      end
+
+      new_thread("same_db_parallel_thread3") do
+        ActiveRecord::Base.on_primary do
+          pause_and_mark_ready
+          result = execute_sql("SELECT 'fast-ish query',SLEEP(#{query_delay.fetch(:medium)})")
+          results.push(result)
+        end
+      end
+
+      wait_for_threads_to_be_ready
+
+      thread_exection_time = Benchmark.realtime do
+        execute_and_wait_for_threads
+      end
+
+      minimum_serial_query_exection_time = query_delay.values.map(&:to_f).sum
+      # Arbitrarily faster time such that there must have been some parallelization
+      max_parallel_time = minimum_serial_query_exection_time - 0.1
+      assert_operator(max_parallel_time, :>, thread_exection_time)
+
+      rows = results.map(&:first)
+      result_strings = rows.map(&:first)
+      # This order cannot be guaranteed but it likely given the artifical delays
+      expect(result_strings).must_equal(
+        [
+          "faster query",
+          "fast-ish query",
+          "slower query"
+        ]
+      )
```

Comment:
```suggestion
      assert_equal([
        "faster query",
        "fast-ish query",
        "slower query"
      ], result_strings)
```

Reply:
Would you prefer all these expectations to be replaced with assertion syntax?